db/change_cache.go

```diff
@@ -878,33 +874,32 @@ func (h *LogPriorityQueue) Pop() interface{} {
 // ////// SKIPPED SEQUENCE QUEUE
 
 func (c *changeCache) RemoveSkipped(x uint64) error {
-	err := c.skippedSeqs.Remove(x)
-	c.db.DbStats.Cache().SkippedSeqLen.Set(int64(c.skippedSeqs.skippedList.Len()))
+	err := c.skippedSeqs.removeSeq(x)
+	c.db.DbStats.Cache().SkippedSeqLen.Set(int64(len(c.skippedSeqs.list)))
+	c.db.DbStats.Cache().SkippedSeqCap.Set(int64(cap(c.skippedSeqs.list)))
+	if err == nil {
+		c.db.DbStats.Cache().NumCurrentSeqsSkipped.Add(-1)
+	}
 	return err
 }
 
 // Removes a set of sequences.  Logs warning on removal error, returns count of successfully removed.
-func (c *changeCache) RemoveSkippedSequences(ctx context.Context, sequences []uint64) (removedCount int64) {
-	numRemoved := c.skippedSeqs.RemoveSequences(ctx, sequences)
-	c.db.DbStats.Cache().SkippedSeqLen.Set(int64(c.skippedSeqs.skippedList.Len()))
-	return numRemoved
+func (c *changeCache) RemoveSkippedSequences(ctx context.Context, sequences []uint64) {
+	// this can be used for removal or ranges, pending CBG-3855
 }
 
 func (c *changeCache) WasSkipped(x uint64) bool {
 	return c.skippedSeqs.Contains(x)
 }
 
-func (c *changeCache) PushSkipped(ctx context.Context, sequence uint64) {
-	err := c.skippedSeqs.Push(&SkippedSequence{seq: sequence, timeAdded: time.Now()})
-	if err != nil {
-		base.InfofCtx(ctx, base.KeyCache, "Error pushing skipped sequence: %d, %v", sequence, err)
-		return
-	}
-	c.db.DbStats.Cache().SkippedSeqLen.Set(int64(c.skippedSeqs.skippedList.Len()))
-}
-
-func (c *changeCache) GetSkippedSequencesOlderThanMaxWait() (oldSequences []uint64) {
-	return c.skippedSeqs.getOlderThan(c.options.CacheSkippedSeqMaxWait)
+func (c *changeCache) PushSkipped(ctx context.Context, startSeq uint64, endSeq uint64) {
+	newEntry := NewSkippedSequenceRangeEntry(startSeq, endSeq)
+	numSequences := newEntry.getNumSequencesInEntry()
+	c.skippedSeqs.PushSkippedSequenceEntry(newEntry)
+	c.db.DbStats.Cache().SkippedSeqLen.Set(int64(len(c.skippedSeqs.list)))
+	c.db.DbStats.Cache().SkippedSeqCap.Set(int64(cap(c.skippedSeqs.list)))
```

Comment:
Updating four stats every time we PushSkipped feels like potentially a lot of overhead since they are all atomic writes.  This is particularly true for SkippedSeqCap which we expect to change infrequently, but even for the others I don't know if we need this kind of granularity.  

One option would be to only update these stats in changeCache.updateStats (which runs once per minute).  That would be fine for sgcollect/stats.log - the question is whether we'd expect a customer to want finer granularity than that, and what the performance overhead would be of updating all these in real time.

Reply:
Completely forgot `changeCache.updateStats` was a thing. Updated to use this!